src/async_traits.rs

```diff
@@ -371,53 +390,86 @@ where
         let async_conn = Connection(Self::as_async_conn(&conn).0.clone());
         match f(async_conn).await {
             Ok(value) => {
-                conn.run_with_shared_connection(|conn| {
-                    Conn::TransactionManager::commit_transaction(conn).map_err(E::from)
-                })
-                .await?;
-                Ok(value)
+                match conn
+                    .run_with_shared_connection(|conn| {
+                        Conn::TransactionManager::commit_transaction(conn)
+                            .map_err(|err| E::from(RunError::User(err)))
+                    })
+                    .await
+                {
+                    Ok(()) => Ok(value),
+                    // XXX: we should try to roll this back
+                    Err(RunError::User(err)) => Err(err),
+                    Err(RunError::RuntimeShutdown) => Err(RunError::RuntimeShutdown.into()),
+                }
             }
             Err(user_error) => {
                 match conn
                     .run_with_shared_connection(|conn| {
-                        Conn::TransactionManager::rollback_transaction(conn).map_err(E::from)
+                        Conn::TransactionManager::rollback_transaction(conn)
+                            .map_err(|err| E::from(RunError::User(err)))
                     })
                     .await
                 {
                     Ok(()) => Err(user_error),
-                    Err(err) => Err(err),
+                    Err(RunError::User(err)) => Err(err),
+                    Err(RunError::RuntimeShutdown) => Err(RunError::RuntimeShutdown.into()),
                 }
             }
         }
     }
 }
 
+fn handle_spawn_blocking_error<T, E>(
+    result: Result<Result<T, E>, JoinError>,
+) -> Result<T, RunError<E>> {
+    match result {
+        Ok(Ok(v)) => Ok(v),
+        Ok(Err(err)) => Err(RunError::User(err)),
+        Err(err) => {
+            if err.is_cancelled() {
+                // The only way a spawn_blocking task can be marked cancelled
+                // is if the runtime started shutting down _before_
+                // spawn_blocking was called.
+                Err(RunError::RuntimeShutdown)
+            } else if err.is_panic() {
+                // Propagate panics.
+                std::panic::panic_any(err.into_panic());
+            } else {
+                // Not possible to reach this as of Tokio 1.40, but maybe in
+                // future versions.
+                panic!("unexpected JoinError: {:?}", err);
+            }
```

Comment:
very minor style suggestion, take it or leave it: i might roll the if conditions into the match, personally:
```suggestion
        // The only way a spawn_blocking task can be marked cancelled
        // is if the runtime started shutting down _before_
        // spawn_blocking was called.
        Err(err) if err.is_cancelled() => Err(RunError::RuntimeShutdown),
        // Propagate panics
        Err(err) if err.is_panic() => std::panic::panic_any(err.into_panic()),
        // Not possible to reach this as of Tokio 1.40, but maybe in
        // future versions.
        Err(err) => panic!("unexpected JoinError: {:?}", err),
```